include/array/array.h

```diff
@@ -2264,12 +2264,39 @@ class array {
    * `alloc`. This is a deep copy of the contents of `other`. */
   array(const array& other, const Alloc& alloc) : array(alloc) { assign(other); }
 
+  /** Move construct from another array `other`. If the allocator of this array
+   * is propagate on container move assignment, this operation moves the allocation
+   * of other to this array, and the other array becomes a default constructed
+   * array. If the allocator of this and the other array are non-equal, each
+   * element is move-constructed into a new allocation. */
+  array(array&& other) : buffer_(nullptr), buffer_size_(0), base_(nullptr) {
+    using std::swap;
+    if (std::allocator_traits<allocator_type>::propagate_on_container_move_assignment::value) {
```

Comment:
There are at least two other nearly identical pieces of code, it would be good to try to figure out how to share them. Here is about equivalent to `operator=(array&& other)` and the else branch is the same as the one in `array(array&& other, const Alloc& alloc)`.

Maybe the best thing to do is just add helper functions for swapping?

Reply:
Sure. std::forward for rescue. PTAL.